com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/TimeProfiler.java

```diff
@@ -0,0 +1,335 @@
+        marks = new ArrayList();
+        RuntimeMXBean bean = ManagementFactory.getRuntimeMXBean();
+        long startTime = bean.getStartTime(); //Returns the start time of the Java virtual machine in milliseconds.
+        buildTime = startTime;
+        rootScope = new ProfilingScope();
+        rootScope.name = "Total time";
+        rootScope.startTime = startTime;
+        currentScope = rootScope;
+
+        ProfilingScope initScope = new ProfilingScope();
+        initScope.name = "Java VM init";
+        initScope.startTime = startTime;
+        initScope.endTime = time();
+        rootScope.children = new ArrayList<ProfilingScope>();
+        rootScope.children.add(initScope);
+        initScope.parent = rootScope;
+
+        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                long reportStartTime = time();
+
+                //Close all unclosed scopes
+                while(currentScope != rootScope) {
+                    addData("forceFinishedScope", true);
+                    addData("color", "#FF0000");
+                    stop();
+                };
+                stop();
+
+                try {
+                    String jsonReport = generateJSON();
+                    String format = fileFormat.getFormat();
+                    String name = fileName.replace(format, "");
+                    File reportFile = new File(reportFolderPath, name + FILENAME_POSTFIX + format);
+                    if (fileFormat == ReportFormat.JSON) {
+                        saveJSON(jsonReport, reportFile);
+                    } else {
+                        saveHTML(jsonReport, reportFile);
+                    }
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+
+                long reportEndTime = time();
+                System.out.printf("\nTime profiler report creation took %.2f seconds", (reportEndTime - reportStartTime)/1000.0f);
+            }
+        }));
+    }
+
+    public static void start(String scopeName) {
+        if (rootScope == null) {
+            return;
+        }
+        if (currentScope.children == null) {
+            currentScope.children = new ArrayList<ProfilingScope>();
+        }
+        ProfilingScope scope = new ProfilingScope();
+        scope.name = scopeName;
+        scope.startTime = time();
+        scope.parent = currentScope;
+        currentScope.children.add(scope);
+        currentScope = scope;
+    }
+
+    public static void startF(String fmt, Object... args) {
+        start(String.format(fmt, args));
+    }
+
+    public static void stop() {
+        if (rootScope == null) {
+            return;
+        }
+        currentScope.endTime = time();
+        currentScope = currentScope.parent;
+    }
+
+    public static void addMark(String shortName, String fullName) {
```

Comment:
This is a way to show a more on the flame chart, just a flag that something happened
like this:
<img width="246" alt="image" src="https://user-images.githubusercontent.com/2209596/167249848-150c5619-cc82-468d-a9e7-93f28f52a4fd.png">